sklearn/mixture/base.py

```diff
@@ -0,0 +1,425 @@
+        """
+        self._check_is_fitted()
+        X = _check_X(X, None, self.n_features)
+        return self._estimate_weighted_log_prob(X).argmax(axis=1)
+
+    def predict_proba(self, X):
+        """Predict posterior probability of data per each component.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        Returns
+        -------
+        resp : array, shape = (n_samples, n_components)
+        """
+        self._check_is_fitted()
+        X = _check_X(X, None, self.n_features)
+        _, resp = self._estimate_log_prob_resp(X)
+        return resp
+
+    def _estimate_weighted_log_prob(self, X):
+        """Estimate the weighted log-probabilities, log P(X | Z) + log weights.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        Returns
+        -------
+        weighted_log_prob : array, shape = (n_features, n_component)
+        """
+        return self._estimate_log_prob(X) + self._estimate_log_weights()
+
+    @abstractmethod
+    def _estimate_log_weights(self):
+        """Estimate log-weights in EM algorithm, E[ log pi ] in VB algorithm.
+
+        Returns
+        -------
+        log_weight : array, shape = (n_components, )
+        """
+        pass
+
+    @abstractmethod
+    def _estimate_log_prob(self, X):
+        """Estimate the log-probabilities log P(X | Z).
+
+        Compute the log-probabilities per each component for each sample.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        Returns
+        -------
+        log_prob : array, shape = (n_samples, n_component)
+        """
+        pass
+
+    def _estimate_log_prob_resp(self, X):
+        """Estimate log probabilities and responsibilities for each sample.
+
+        Compute the log probabilities, weighted log probabilities per
+        component and responsibilities for each sample in X with respect to
+        the current state of the model.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        Returns
+        -------
+        log_prob_norm : array, shape = (n_samples,)
+            log p(X)
+
+        log_prob : array, shape = (n_samples, n_components)
+            log p(X|Z) + log weights
+
+        responsibilities : array, shape = (n_samples, n_components)
+        """
+        weighted_log_prob = self._estimate_weighted_log_prob(X)
+        log_prob_norm = logsumexp(weighted_log_prob, axis=1)
+        with np.errstate(under='ignore'):
+            # ignore underflow
+            resp = np.exp(weighted_log_prob - log_prob_norm[:, np.newaxis])
+        return log_prob_norm, resp
+
+    def _print_verbose_msg_init_beg(self, n_init):
+        """Print verbose message on initialization."""
+        if self.verbose == 1:
+            print("Initialization %d" % n_init)
+        elif self.verbose == 2:
```

Comment:
You should probably use `self.verbose >= 2`, as it is tempting to set `verbose=10` to have full verbosity without checking what is the highest value used in the code.